server/src/test/java/org/elasticsearch/cluster/coordination/JoinTaskExecutorTests.java

```diff
@@ -488,6 +499,92 @@ public void testIgnoresOlderTerms() throws Exception {
+
+        final var joinTaskExecutor = new JoinTaskExecutor(allocationService, rerouteService);
+
+        final var actualizedDesiredNodes = randomList(0, 5, this::createActualizedDesiredNode);
+        final var pendingDesiredNodes = randomList(0, 5, this::createPendingDesiredNode);
+        final var joiningDesiredNodes = randomList(1, 5, this::createPendingDesiredNode);
+
+        final var masterNode = newDiscoveryNode("master");
+        final var discoveryNodes = DiscoveryNodes.builder()
+            .add(masterNode)
+            .localNodeId(masterNode.getId())
+            .masterNodeId(masterNode.getId());
+
+        for (DesiredNodeWithStatus actualizedDesiredNode : actualizedDesiredNodes) {
+            discoveryNodes.add(newDiscoveryNode(actualizedDesiredNode.externalId()));
+        }
+
+        final List<DiscoveryNode> joiningNodes = joiningDesiredNodes.stream()
+            .map(desiredNode -> newDiscoveryNode(desiredNode.externalId()))
+            .toList();
+
+        // Add some nodes in the cluster that are not part of the desired nodes
+        for (int i = 0; i < randomInt(5); i++) {
+            discoveryNodes.add(newDiscoveryNode(UUIDs.randomBase64UUID(random())));
+        }
+
+        final var desiredNodes = DesiredNodes.create(
+            randomAlphaOfLength(10),
+            randomInt(10),
+            concatLists(actualizedDesiredNodes, pendingDesiredNodes, joiningDesiredNodes)
+        );
+
+        var clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(discoveryNodes)
+            .metadata(Metadata.builder().putCustom(DesiredNodesMetadata.TYPE, new DesiredNodesMetadata(desiredNodes)))
+            .build();
+
+        var tasks = joiningNodes.stream().map(node -> JoinTask.singleNode(node, "join", NOT_COMPLETED_LISTENER, 0L)).toList();
+
+        clusterState = ClusterStateTaskExecutorUtils.executeAndAssertSuccessful(clusterState, joinTaskExecutor, tasks);
+
+        final var updatedDesiredNodes = DesiredNodes.latestFromClusterState(clusterState);
+        assertThat(updatedDesiredNodes, is(notNullValue()));
+
+        assertThat(updatedDesiredNodes.nodes(), hasSize(desiredNodes.nodes().size()));
+        assertDesiredNodesStatusIsCorrect(
+            clusterState,
+            Stream.concat(actualizedDesiredNodes.stream(), joiningDesiredNodes.stream()).map(DesiredNodeWithStatus::desiredNode).toList(),
+            pendingDesiredNodes.stream().map(DesiredNodeWithStatus::desiredNode).toList()
+        );
+    }
+
+    @SafeVarargs
+    private List<DesiredNodeWithStatus> concatLists(List<DesiredNodeWithStatus>... nodeLists) {
+        assertThat(nodeLists.length, is(greaterThan(0)));
+
+        List<DesiredNodeWithStatus> concatList = new ArrayList<>();
+        for (List<DesiredNodeWithStatus> desiredNodeWithStatuses : nodeLists) {
+            concatList.addAll(desiredNodeWithStatuses);
+        }
+
+        return Collections.unmodifiableList(concatList);
```

Comment:
I think this can be simplified to `StreamSupport.stream(nodeLists).flatMap(List::stream).toList()`.